### PR TITLE
feat: check class & transaction hashes

### LINKED ISSUE #2346: P2P sync: use the proto's class_hash & transaction_hash fields to bail early
Now:
- we compute the hashes and then
- we verify commitments, which is the point where we can bail on mismatch.

With the class_hash and transaction_hash fields we can:
- compute the hashes and immediately bail on hash mismatch,
- this way avoiding even heavier computation which will fail.

### PR DESCRIPTION
**Pull request by vbar**
for equality with locally-computed hashes (and fail on mismatch).

Fixes #2346.

### CODE REVIEW

File: crates/common/src/class_definition.rs [hunk 1]
```diff
@@ -8,7 +8,13 @@ use serde::{Deserialize, Serialize};
 use serde_json::value::RawValue;
 use serde_with::serde_as;
 
-use crate::{ByteCodeOffset, EntryPoint};
+use crate::{ByteCodeOffset, ClassHash, EntryPoint};
+
+#[derive(Debug)]
+pub struct ClassDefinitionWithHash<'a> {
+    pub cls_def: ClassDefinition<'a>,
```
Comment (kkovaacs) [inline crates/common/src/class_definition.rs:15]: Just a small nit: can we maybe call this `definition` instead?

**Review by kkovaacs — APPROVED** [review]
LGTM, nice work!

File: crates/pathfinder/src/sync/class_definitions.rs [hunk 2]
```diff
@@ -222,11 +235,16 @@ fn compute_hash_impl(input: ClassWithLayout) -> anyhow::Result<Class> {
...
     }?;
 
-    Ok(Class {
-        block_number,
-        definition,
-        hash,
-    })
+    if hash != layout.hash {
+        tracing::debug!(input_hash=%layout.hash, actual_hash=%hash, "Class hash mismatch");
+        Err(SyncError2::ClassHashComputationError.into())
```
Comment (CHr15F0x) [inline crates/pathfinder/src/sync/class_definitions.rs:240]: nit: I wonder if we should differentiate between a mismatch (ie. proper layout, bad values) and hash computation error (which means bad layout).
So for the former: ClassHashMismatch (maybe we need to add this variant)
and for the latter: ClassHashComputationError (which really is an InvalidDto but at a higher level)

**Review by CHr15F0x — APPROVED** [review]
LGTM :bow: 

A small nit - maybe we should use 2 variants there to differentiate between utter garbage (invalid dto, resulting in class computation error) and a mismatch (valid dto, the values dont match).

File: crates/pathfinder/src/sync/class_definitions.rs (hunk 2 above)
Comment (vbar) [reply crates/pathfinder/src/sync/class_definitions.rs]: Well, I wasn't keen on extending SyncError / SyncError2, b/c

File: crates/common/src/class_definition.rs (hunk 1 above)
Comment (vbar) [reply crates/common/src/class_definition.rs]: Upon reflection, I don't like the whole `ClassDefinitionWithHash` - now that I've finally fit the carpet, :-) it isn't really used that much...

**Pull request merged**